stan/math/prim/mat/fun/matrix_power.hpp

```diff
@@ -0,0 +1,57 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_MATRIX_POWER_HPP
+#define STAN_MATH_PRIM_MAT_FUN_MATRIX_POWER_HPP
+
+#include <stan/math/prim/scal/err/invalid_argument.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/err/check_finite.hpp>
+#include <stan/math/prim/mat/err/check_square.hpp>
+
+namespace stan {
+namespace math {
+/**
+ * Returns the nth power of the specific matrix. M^n = M * M * ... * M.
+ *
+ * @tparam T type of elements in the matrix
+ * @tparam R number of rows in matrix
+ * @tparam C number of columns in matrix
+ * @param[in] M a square matrix
+ * @param[in] n exponent
+ * @return nth power of M
+ * @throw std::domain_error if the matrix contains NaNs or infinities.
+ * @throw std::invalid_argument if the exponent is negative or the matrix is not
+ * square.
+ */
+template <typename T, int R, int C>
+inline Eigen::Matrix<T, R, C> matrix_power(const Eigen::Matrix<T, R, C> &M,
+                                           const int n) {
+  check_square("matrix_power", "M", M);
+  if (n < 0)
+    invalid_argument("matrix_power", "n", n, "is ", ", but must be >= 0!");
+  if (M.rows() == 0)
+    invalid_argument("matrix_power", "M.rows()", M.rows(), "is ",
+                     ", but must be > 0!");
+  check_finite("matrix_power", "M", M);
+  Eigen::Matrix<T, R, C> result
+      = Eigen::Matrix<T, R, C>::Identity(M.rows(), M.cols());
+  int nn = n;
```

Comment:
I'm not going to review this, just leave a couple comments.

This can be a for loop, which keeps scope cleaner:

```
for (int nn = n;  nn > 0; nn /= 2) {
 ...
}
```

Also, you can start with a single matrix rather than the identity to avoid the identity multiplication (and return identityin special case that power is 0).

Reply:
Loop cleaned up.